gnsi/acltrie/trie.go

```diff
@@ -152,79 +165,104 @@ func (t *Trie) Insert(r *pathzpb.AuthorizationRule) error {
+func (t *Trie) Probe(path *gpb.Path, user string, mode pathzpb.Mode) pathzpb.Action {
+	potentialPolicies := []*trieNode{}
+	maxLen := 0
+
+	// Walk the matching rules, keeping only the longest ones that contain either the user or a group that the user belongs too.
+	t.walk(func(node *trieNode, walkPath *gpb.Path) (bool, error) {
+		if res := util.ComparePaths(path, walkPath); res != util.Equal && res != util.Subset {
+			return false, nil
+		}
+		if !node.hasPolicy {
+			return true, nil
+		}
+		if act, _ := node.getAction(user, mode, t.memberships); act != pathzpb.Action_ACTION_UNSPECIFIED {
+			if len(walkPath.Elem) > maxLen {
+				potentialPolicies = nil
+				maxLen = len(walkPath.Elem) - 1
+			}
+			potentialPolicies = append(potentialPolicies, node)
+		}
+
+		return true, nil
+	})
+	if len(potentialPolicies) == 0 {
+		return pathzpb.Action_ACTION_DENY
+	}
+
+	// Pick the policies with the largest number of definite keys.
+	var bestPaths []*trieNode
+	maxSetKeys := -1
+	for _, p := range potentialPolicies {
+		path := &gpb.Path{
+			Elem: make([]*gpb.PathElem, maxLen),
+		}
+		n := p
+		setKey := 0
+		for i := maxLen - 1; i >= 0; i-- {
+			path.Elem[i] = n.elem
+			setKey += setKeys(path.Elem[i].Key)
+			n = p.parent
+		}
+		if setKey >= maxSetKeys {
+			maxSetKeys = setKey
+			bestPaths = append(bestPaths, p)
+		}
+	}
+
+	var finalAction pathzpb.Action
+	// Prefer user over groups and DENY over permit.
+	for _, n := range bestPaths {
+		act, isUser := n.getAction(user, mode, t.memberships)
+		switch {
+		case isUser && act == pathzpb.Action_ACTION_DENY: // Prefer user and deny, so return immediately.
+			return act
+		case isUser: // Prefer a user over group.
+			finalAction = act
+		case finalAction != pathzpb.Action_ACTION_DENY: // Prefer deny over allow.
+			finalAction = act
+		}
+	}
+
+	return finalAction
+}
+
+// walk explores the trie in depth first order and invokes walkFn on every node.
 // To continue exploring children of the node, the walkFn must return true.
-func (t *Trie) walk(walkFn func(node *trieNode, depth int) (bool, error)) error {
+// Note: the path object is modified between calls.
+func (t *Trie) walk(walkFn func(*trieNode, *gpb.Path) (bool, error)) error {
 	type traversalNode struct {
 		node  *trieNode
 		depth int
 	}
 
-	queue := []*traversalNode{{node: t.root}}
-	for len(queue) > 0 {
-		front := queue[0]
-		queue = queue[1:]
-		for _, c := range front.node.children {
-			cont, err := walkFn(c, front.depth)
+	path := &gpb.Path{}
+	stack := []*traversalNode{{node: t.root}}
+
+	for len(stack) > 0 {
+		last := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+		for _, c := range last.node.children {
+			// Handle inconsistency with the start of a path being at depth 1,
+			// but should be 0th elem in the slice.
+			if last.depth == 1 {
+				path.Elem = []*gpb.PathElem{last.node.elem}
+			}
+			path.Elem = append(path.Elem[:last.depth], c.elem)
+			cont, err := walkFn(c, path)
```

Comment:
I believe the reason for the inconsistency is due to the `walkFn` being invoked for the children *before* they're added to the stack as opposed to *directly after* popping from the stack. This is causing the nodes to be visited in a non-DFS order, which is what is causing the issue. If `walkFn` is called prior to the inner loop, then the visiting will occur in DFS order, and then the path would be consistent.

An implication is that now the root node will be visited, and the current algorithm doesn't visit the root.